### PR TITLE
Improve Sheaf's performance, fix bugs

### PR DESCRIPTION
**Pull request by Monkatraz**
This PR is mostly focusing on Sheaf's performance. This involved a lot of different changes. Sorry for the big PR.

Changes:
* Fix CodeMirror package duplication issues - required shameful dependency hoisting and manual `resolve.dedupe` configuration
* Fixed Sheaf's live preview not actually being live
* Migrate all workers to a new, slicker Comlink based system that is much simpler to use and much smaller than Threads
* Add a basic module worker polyfill so that _some_ workers kind of work on Firefox in development mode. This is probably temporary as Firefox is seemingly going to be getting module workers soon.
* Fix formatting issues with `<pre>` blocks in Svelte components
* Updated the `cm.ts` CodeMirror exports to be up to date
* Extended the `@wikijump/util` `animationFrame()` and `idleCallback()` methods to take executor functions
* Added a good `requestIdleCallback` polyfill that actually provides some performance benefit on Safari
* Updated the `@wikijump/ftml-components` code and math blocks to render on an idle callback
* Massively improve Sheaf live preview performance via CSS containment changes
* Added a super fast string UTF-8 byte length function to `@wikijump/util`, which is used to very quickly get the text size of Sheaf's contents
* Switch to using the tiny `micromorph` library for live preview morphing
* Moved the Prism highlighter off the main thread - now the library will only be downloaded if a code block has to be highlighted, which will be done asynchronously
* Updated ftml-wasm bundle
* Fixed a bunch of misc. bugs

Next PR will probably be me fixing a bunch of issues that are cropping up with production builds of the Vite stuff.

### CODE REVIEW

File: web/modules/cm-espells/src/MisspellingTooltip.svelte [hunk 1]
```diff
@@ -102,9 +102,9 @@
 
   <hr />
 
-  <pre class="code cm-spellcheck-tip-source">
-      <code>{$t("#-source", { slice: word.word, from: word.from, to: word.to })}</code>
-  </pre>
+  <pre class="code cm-spellcheck-tip-source"><code
+      >{$t("#-source", { slice: word.word, from: word.from, to: word.to })}</code
+    ></pre>
```
Comment (emmiegit) [inline web/modules/cm-espells/src/MisspellingTooltip.svelte:107]: lol, clever workaround to avoid whitespace

File: web/modules/cm-espells/src/espells/interface.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+  protected async createWorker() {
+    if (DICTIONARIES.hasOwnProperty(this.locale)) {
+      const { aff: affURL, dic: dicURL } = await DICTIONARIES[this.locale]()
+
+      // we can't use the static fromURL method on Espells because of Comlink limitations
+      // so it has to be reimplemented basically - we have to resolve the urls
+
+      const aff = await fetch(affURL).then(res => res.text())
+      const dic = Array.isArray(dicURL)
+        ? await Promise.all(dicURL.map(dic => fetch(dic))).then(res =>
```
Comment (emmiegit) [inline web/modules/cm-espells/src/espells/interface.ts]: This is a bit difficult to read, maybe worth adding a comment or two explaining what the parts of the callbacks are doing?

File: web/modules/comlink/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+   * automatically handles binding methods and worker creation.
+   *
+   * @param methods - The methods to bind to the class instance, which when
+   *   called will be passed the worker instance.
+   */
+  static of<T>(
+    methods: (keyof T)[]
+  ): abstract new () => AbstractWorkerBase<T> & Comlink.RemoteObject<T> {
+    // @ts-ignore
+    return class extends AbstractWorkerBase<T> {
```
Comment (emmiegit) [inline web/modules/comlink/src/index.ts:56]: You can do this? Incredible

File: web/modules/sheaf/src/components/Wikitext.svelte [hunk 4]
```diff
@@ -70,53 +79,54 @@
...
+        element.innerText = ""
+        element.appendChild(fragment)
+      }
+    })
+
+    await animationFrame(() => {
+      // prepend style with a index comment so that each style string is unique
+      stylesheets = styles.map(
+        (style, idx) => `\n/* stylesheet ${idx + 1} */\n\n${style}\n`
+      )
```
Comment (emmiegit) [inline web/modules/sheaf/src/components/Wikitext.svelte:127]: Interesting, why is this necessary?

File: web/modules/sheaf/src/core.ts [hunk 5]
```diff
@@ -72,7 +72,8 @@ export class SheafCore {
   }
 
   private update(update: ViewUpdate) {
-    if (!update.docChanged && !update.selectionSet) return
+    // if (!update.docChanged && !update.selectionSet) return
```
Comment (emmiegit) [inline web/modules/sheaf/src/core.ts:77]: Why is this code commented out?

**Review by emmiegit — APPROVED** [review]

File: web/modules/cm-espells/src/MisspellingTooltip.svelte (hunk 1 above)
Comment (Monkatraz) [reply web/modules/cm-espells/src/MisspellingTooltip.svelte:107]: Yeah, that's what Prettier does. It is a cool workaround.

File: web/modules/comlink/src/index.ts (hunk 3 above)
Comment (Monkatraz) [reply web/modules/comlink/src/index.ts:56]: yeah honestly when I got this horrifying thing working I was very surprised

File: web/modules/cm-espells/src/espells/interface.ts (hunk 2 above)
Comment (Monkatraz) [reply web/modules/cm-espells/src/espells/interface.ts]: Yeah, or I should maybe rewrite it. It's just trying to be fast by doing all the fetch requests in parallel.

File: web/modules/sheaf/src/components/Wikitext.svelte (hunk 4 above)
Comment (Monkatraz) [reply web/modules/sheaf/src/components/Wikitext.svelte:127]: The stylesheets are uniquely identified by their contents. What is uniquely identifying them is Svelte - below in the markup there is a `{#each}` block that is putting the stylesheet strings into the `<head>`. When the stylesheets array updates, it compares the previous list to the new one.

If you had two stylesheets with the same contents, it would confuse Svelte.

File: web/modules/sheaf/src/core.ts (hunk 5 above)
Comment (Monkatraz) [reply web/modules/sheaf/src/core.ts:77]: Ah, oops. That's temporary. With the original check, moving your cursor at all would cause the FTML to get rerendered. For now, the editor state will only update if the document changes.

Later, that will need to be properly fixed, because if we add sourcemapping there will need to be a bunch of editor state that updates with scrolling, active lines, etc.

**Comment by Monkatraz** [comment]
thanks @ammongit
